oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    private String getBaseIndexPath(String path) {
+        String trimmedPath = PurgeOldVersionUtils.trimSlash(path);
+        String[] fragmentedPath = trimmedPath.split("/");
+        String indexName = fragmentedPath[fragmentedPath.length - 1];
+        String baseIndexName = PurgeOldVersionUtils.getBaseIndexName(indexName);
+        return stringJoiner(fragmentedPath,
+                fragmentedPath.length - 2, "/")
+                .append(baseIndexName).toString();
+    }
+
+    private Set<String> getIndexVersionNames(Set<String> versionedIndexPaths) {
+        Set<String> indexVersionNames = new HashSet<>();
+        for (String indexPath : versionedIndexPaths) {
+            String trimmedIndexPath = PurgeOldVersionUtils.trimSlash(indexPath);
+            String[] splitPathArray = trimmedIndexPath.split("/");
+            String indexName = splitPathArray[splitPathArray.length - 1];
+            indexVersionNames.add(indexName);
+        }
+        return indexVersionNames;
+    }
+
+    private List<IndexName> getIndexNameObjectList(Set<String> versionedIndexPaths) {
+        Set<String> allIndexNameVersions = getIndexVersionNames(versionedIndexPaths);
+        List<IndexName> indexNameObjectList = getVersionedIndexNameObjects(allIndexNameVersions);
+        return indexNameObjectList;
+    }
+
+    private void purgeOldIndexVersion(NodeStore store,
+                                      List<IndexVersionOperation> toDeleteIndexNameObjectList, String trimmedIndexPath) throws CommitFailedException {
+        String indexPathParent = "/" + trimmedIndexPath.substring(0, trimmedIndexPath.lastIndexOf('/'));
+        for (IndexVersionOperation toDeleteIndexNameObject : toDeleteIndexNameObjectList) {
+            NodeState root = store.getRoot();
+            NodeBuilder rootBuilder = root.builder();
+            NodeBuilder nodeBuilder = PurgeOldVersionUtils.getNode(rootBuilder, indexPathParent + "/" + toDeleteIndexNameObject.getIndexName().getNodeName());
+            if (nodeBuilder.exists()) {
+                if (toDeleteIndexNameObject.getOperation() == IndexVersionOperation.Operation.DELETE_HIDDEN_AND_DISABLE) {
+                    nodeBuilder.setProperty("type", "disabled", Type.STRING);
+                    PurgeOldVersionUtils.recursiveDeleteHiddenChildNodes(store, trimmedIndexPath.substring(0, trimmedIndexPath.lastIndexOf('/')) + "/" + toDeleteIndexNameObject.getIndexName().getNodeName());
+                } else if (toDeleteIndexNameObject.getOperation() == IndexVersionOperation.Operation.DELETE) {
+                    nodeBuilder.remove();
+                }
+                EditorHook hook = new EditorHook(
+                        new IndexUpdateProvider(new PropertyIndexEditorProvider()));
+                store.merge(rootBuilder, hook, CommitInfo.EMPTY);
+            } else {
+                log.error("nodebuilder null for path " + indexPathParent + "/" + toDeleteIndexNameObject.getIndexName().getNodeName());
```

Comment:
I think it's a bit weird to log an error, but not throw an exception. I would throw an exception, catch it at the caller, and then log an error there. To avoid having a lot of code indented, I would use:

    if (!nodeBuilder.exists()) {
        throw ...
    }
    ... code for normal case ...